36 checkpermission middleware

sanjaysah101 opened the pull request:
- Implement RBAC and improve security
- Refactor AuthController, AuthService
- Implement checkPermission middleware
- Rename `checkBearerToken` to `extractTokenFromBearerString` for clarity
- Add comprehensive unit tests for the extractTokenFromBearerString function (#39)
- Improve error handling for missing or invalid bearer tokens

sanjaysah101 committed:
feat(auth): implement RBAC and improve security

- Add checkPermission middleware for role-based access control
- Rename validateMongoId to objectIdSchema for clarity
- Implement validateObjectId function for ID validation
- Enhance security in auth.ts for access token checks
- Add FORBIDDEN error type to error handling
- Fix references in Mongoose schemas
- Update Zod schemas for improved validation
- Correct updatedBy field in permission and roles models

This commit introduces role-based access control and makes several
improvements to enhance security and data validation across the
authentication system.

sanjaysah101 committed:
refactor(auth): code optimization

sanjaysah101 committed:
perf(middleware): optimize checkPermission with aggregation pipeline

- Implement getUserRolesAndPermissionsByUserId using MongoDB aggregation
- Improve permission checking efficiency for role-based access control
- Reduce database queries and enhance overall performance

ansopedia commented on src/api/v1/permission/permission.model.ts:
```diff
@@ -24,8 +24,24 @@ const PermissionSchema: Schema<Permission> = new Schema<Permission>(
     },
-    createdBy: Types.ObjectId,
-    updatedBy: Types.ObjectId,
+    createdBy: {
+      type: String,
+      required: true,
+      validate: {
+        validator: (v: string) => Types.ObjectId.isValid(v),
+        message: 'createdBy must be a valid MongoDB ObjectId string',
+      },
+      ref: 'User',
+    },
+    updatedBy: {
+      type: String,
+      required: true,
+      validate: {
+        validator: (v: string) => Types.ObjectId.isValid(v),
+        message: 'updatedBy must be a valid MongoDB ObjectId string',
+      },
+      ref: 'User',
+    },
```
@sanjaysah101, we can simplify the system by removing the `createdBy` and `updatedBy` fields for roles and permissions. Since we'll be using static roles and permissions defined as constants, there's no need to create new roles or permissions from the admin panel. This approach reduces complexity and ensures a more robust service. The admin panel will only allow modifications to the relationships between roles and permissions, and between users and roles. By keeping predefined roles and permissions in the database, we can maintain a clear and predictable flow in the application. On the client side, we can dynamically adjust behaviour when permissions are assigned or removed from roles.

sanjaysah101 replied:
Hi @ansopedia, that sounds great! I think we can remove the CRUD operations for permissions, but we should keep them for `roles`, `rolePermissions`, and `userPermissions`.

For example, allowing the admin to create and manage roles from the admin panel and assign predefined permissions will help us maintain flexibility in our RBAC system. With this setup, the API flow remains predictable, as routes are protected by specific permissions, and users have roles tied to those permissions. If we need to adjust access in the future, we can modify the role or its permissions, ensuring easy and efficient management.

ansopedia commented on src/api/v1/permission/permission.model.ts:
```diff
@@ -24,8 +24,24 @@ const PermissionSchema: Schema<Permission> = new Schema<Permission>(
+      },
+      ref: 'User',
+    },
     isDeleted: {
       type: Boolean,
       default: false,
```
Also, remove the `isDeleted` field since we don't require this any more because it adds extra work.

sanjaysah101 committed:
refactor(auth): rename checkBearerToken and add unit tests

- Rename checkBearerToken to extractTokenFromBearerString for clarity
- Add comprehensive unit tests for extractTokenFromBearerString function
- Improve error handling for missing or invalid bearer tokens
